[Feature/withdrawal] 탈퇴한 계정을 복구하는 API

c0z-y-fiction opened the pull request:
## ✅ PR 요약
- 관련 이슈 번호: #64 
- 작업 요약: 회원 탈퇴를 복구하는 API 구현(이메일로 받은 코드를 유저가 입력하면 해당 코드가 유효한지 검증, 해당 유저의 탈퇴를 번복하고 계정을 활성화).

## 📄 상세 내용
- [x] serializer: 복구 코드 입력 검증.
- [x] service: 이메일/인증 코드 검증, 복구 성공 시 계정 활성화 및 탈퇴 요청 삭제(비즈니스 로직).
- [x] view: 탈퇴 계정 복구 요청. 이메일/인증 코드 검증, 복구 성공 시 계정 활성화 및 탈퇴 요청 삭제(HTTP 처리).
- [x] urls 추가, 테스트 코드 작성 및 인증/양식 오류 수정.
- [x] 파일명을 더욱 직관적으로 변경.

## 📸 스크린샷 (선택)

## 📝 기타 참고 사항

## 🧪 PR Checklist
- [x] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [x] 변경 사항에 대한 테스트를 했습니다(버그 수정/기능에 대한 테스트).

c0z-y-fiction committed:
💡 chore: 주석 추가

c0z-y-fiction committed:
✨ feat: 회원 탈퇴 신청했던 유저의 계정을 복구하기 위한 serializer 추가

이메일로 받은 복구 코드를 검증함(AccountRecoverySerializer)

c0z-y-fiction committed:
✨ feat: 회원 탈퇴 신청했던 유저의 계정을 복구하기 위한 service 추가

계정 복구 요청을 받으면 이메일로 받은 인증 코드를 검증하고, 유저 계정을 활성화한 뒤 탈퇴 요청을 삭제. 이에 따라 user.is_active()를 명시적으로 작성.

c0z-y-fiction committed:
✨ feat: 회원 탈퇴 신청했던 유저의 계정을 복구하기 위한 view 추가

(작성 중) 사용자 요청으로 탈퇴했던 계정을 복구할 수 있는 API 추가

c0z-y-fiction committed:
✨ feat: 탈퇴 계정 복구 API 구현 및 service와의 연결

AccountRecoveryAPIView 추가(이메일/인증 코드 검증, 계정 활성화), try/except 성공/실패 처리.

c0z-y-fiction committed:
✨ feat: recover_account() 개선

탈퇴 요청이 없는 계정이 복구를 시도하려고 하면 None 대신 ObjectDoesNotExist 예외를 발생시키도록 변경(추후 오류 처리를 위해), view로 코드 이전

c0z-y-fiction committed:
💡 chore: 계정 복구 관련 url 추가(auth/recover)

c0z-y-fiction committed:
🎨 style: black, isort 코드 포매터 실행

c0z-y-fiction committed:
✨ feat: UserRecoveryJWTAPITest 추가

setUpTestData에서는 user만 생성+setUp에서 URL 체크(mypy 에러 문제). 인증 코드 설정 로직(_set_verification_code) 구현. 복구 요청 시 인증 코드 검증 및 계정 활성화 여부 확인. 계정 복구 후 탈퇴 요청을 삭제.

c0z-y-fiction committed:
🎨 style: black, isort 코드 포매터 실행

c0z-y-fiction committed:
💡 chore: ModelSerializer를 Serializer로 변경

c0z-y-fiction committed:
💡 chore: 타입 힌트, 주석, 오류 타입 변경

ObjectDoesNotExist를 Withdrawals.DoesNotExist로 변경(Django의 스타일을 따라 명시적인 예외 처리).

c0z-y-fiction committed:
✨ feat: AllowAny 권한 추가

탈퇴 신청 유저는 계정으로 자신을 인증할 수 없으므로 AllowAny 권한을 추가하되 이메일+인증 코드로 보안 유지.

c0z-y-fiction committed:
✨ feat: email 필드 추가

c0z-y-fiction committed:
💡 chore: 필요 없는 주석/공백 제거

c0z-y-fiction committed:
✨ feat: UserRecoveryJWTAPITest(회원 탈퇴 복구) 테스트 코드 작성 및 수정

복구 요청은 비인증 상태에서도 진행할 수 있도록 변경. 테스트 도중 회원 계정 복구 HTTP 코드를 확인하는 부분에서 400/401 에러가 번갈아가면서 뜨는 문제 해결(전자: cache_key 양식 변경/후자: force_authenticate 호출 순서 및 타이밍 조정).

c0z-y-fiction committed:
💡 chore: 파일명 변경 재커밋

c0z-y-fiction committed:
💡 chore: import한 함수명 변경에 따른 수정

Umdoong commented on apps/users/services/withdrawals_service.py:
```diff
@@ -0,0 +1,44 @@
+    return withdrawal
+
+
+def recover_account(email: str, verification_code: str) -> User:
```
서비스 함수 내에서 verification_code를 활용하고 있지 않기 때문에 매개변수에서 빼주셔야 합니다.

c0z-y-fiction replied:
이걸 제거하니까 테스트 코드에서 문제가 생겨서 그냥 넣는 게 좋을 것 같습니다. → `400 {'detail': "알 수 없는 오류가 발생했습니다. recover_account() got an unexpected keyword argument 'verification_code'."}`
아니면 제가 모르는 다른 해결 방법이 있을까요?

Umdoong commented on apps/users/views/withdrawals_view.py:
```diff
@@ -0,0 +1,65 @@
+            email_service.verify_code(purpose, **serializer.validated_data)  # 검증 코드 호출
+
+            # 4) 계정 복구
+            user = recover_account(**serializer.validated_data)
```
현 코드에서 user라는 변수에 할당하는 행위는 의미없어보입니다.

Umdoong commented on apps/users/services/withdrawals_service.py:
```diff
@@ -0,0 +1,44 @@
+    # 3) 탈퇴 요청 삭제: withdrawal 모델에서 해당 user를 삭제
+    withdrawal.delete()
+
+    return user
```
user 인스턴스를 활용할 계획이 있으신 게 아니라면 return하지 않아도 될 것 같습니다.

Umdoong commented on apps/users/views/withdrawals_view.py:
```diff
@@ -0,0 +1,65 @@
+        serializer = EmailVerifyCodeSerializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+
+        # 2) 저장된 캐시 불러오기(enum.py)
```
주석이 잘못되었습니다. 저장된 캐시를 불러오는 코드가 아닌 것 같습니다.

Umdoong commented on apps/users/services/withdrawals_service.py:
```diff
@@ -0,0 +1,44 @@
+    return withdrawal
+
+
+def recover_account(email: str, verification_code: str) -> User:
```
해당 서비스를 트랜젝션으로 묶으면 더 좋을 것 같습니다.

Umdoong commented on apps/users/views/withdrawals_view.py:
```diff
@@ -0,0 +1,65 @@
+
+        try:
+            # 3) 인증 코드 검증
+            email_service.verify_code(purpose, **serializer.validated_data)  # 검증 코드 호출
```
verify_code 서비스 내에서 exception handler를 사용하고 있는데 그에 맞춰서 view쪽에서 예외처리를 해주셔야합니다.

Umdoong commented on apps/users/views/withdrawals_view.py:
```diff
@@ -0,0 +1,65 @@
+            return Response({"detail": "계정이 복구되었습니다. 이제 로그인할 수 있습니다."}, status=status.HTTP_200_OK)
+
+        # 6) 실패 응답(오류 제어는 나중에)
+        except Exception as e:
```
서비스 로직에서 발생하는 raise를 디테일하게 except 해주셔야합니다.

c0z-y-fiction replied:
1) 이메일 발송 시스템 문제(500)/이메일 인증 코드 불일치 문제(400?)/해당 이메일로 요청한 탈퇴 요청이 없는 문제(404)까지는 넣었는데, 원래 있던 Exception은 지울까요 아니면 남길까요?
2) 발견하지 못한 문제도 더 찾아서 except에 추가해야 할까요?

Umdoong replied:
raise 시켰을 때 밖에서 모든 Exception을 except해서 400으로 처리하고 있기 때문에 문제가 되는 겁니다!

Umdoong replied:
raise된 것들을 밖에서 하나씩 캐치해서 처리해주시고 그 외의 것들을 일단 Exception으로 해서 500을 반환시키고 logging하는 것도 방법일 것 같아요

Umdoong commented on apps/users/tests/test_withdrawals.py:
```diff
@@ -47,6 +62,59 @@ def test_duplicate_withdrawal_request(self) -> None:
+    # * 탈퇴 복구
+    def test_account_recovery_request(self) -> None:
+        # 1) 탈퇴 요청을 생성
+        self._test_successful_withdrawal_request()
```
테스트 안에서 다른 테스트를 호출하는 것은 이상해보입니다.
Withdrawals.objects.create를 활용하여 데이터를 추가하는 방식이 더 좋아보입니다.

c0z-y-fiction committed:
💡 chore: PR 리뷰대로 service 수정

서비스 함수에서 verification_code 매개 변수 제거, user 인스턴스 활용 계획 없으므로 return 제거, 트랜잭션 추가.

c0z-y-fiction committed:
💡 chore: 테스트 안에서 테스트 호출하는 부분 제거

c0z-y-fiction committed:
✨ feat: except 제어 확장

하나로 묶어서 처리했던 except를 각 상황에 맞게 다시 지정(400/404/500). 기존 Exception은 이외 상황에 대응하기 위해 유지.

c0z-y-fiction committed:
💡 chore: verification_code 복구

없으면 테스트 코드가 작동하지 않습니다. 디버깅 결과: 400 {'detail': "알 수 없는 오류가 발생했습니다. recover_account() got an unexpected keyword argument 'verification_code'."}
